standard/expressions.md

```diff
@@ -6195,6 +6212,55 @@ When a property or indexer declared in a *struct_type* is the target of an assig
 >
 > *end example*
 
+### §ref-assignment-new-clause Ref assignment
+
+The left operand shall be an expression that binds to a ref local variable, a ref parameter (other than `this`), or an out parameter. The right operand shall be an expression that yields a *variable_reference* designating a value of the same type as the left operand.
+
+The right operand shall be definitely assigned at the point of the ref assignment.
+
+When the left operand binds to an `out` parameter, it is an error if that `out` parameter has not been definitely assigned at the beginning of the ref assignment operator.
+
+It is a compile time error if the *ref_safe_scope* of the left operand is wider than the *ref_safe_scope* of the left operand (§ref-span-safety-escape-scopes).
+
+If the left operand is a writeable ref (i.e., it designates anything other than a `ref readonly` local or  `in` parameter), then the right operand shall be a writeable *variable_reference*. If the right operand variable is writeable, the left operand may be declared `ref` or `ref readonly`.
+
+The operation makes the left operand an alias of the right operand variable. The alias may be made read-only even if the right operand variable is writeable.
+
+The ref assignment operator yields a *variable_reference* of the assigned type. It is writeable if the left operand is writeable.
+
+The ref assignment operator shall not evaluate the value of right hand operand, only its reference.
```

Comment:
I think this needs more precision. Consider:

```csharp
static void Main()
{
    int a = 0;
    ref int b = ref M(ref a);
}

static ref int M(ref int y) => ref y;
```

Here the second line of `Main` *does* evaluate the right operand, in that it calls the method. It doesn't then evaluate the variable reference return by `M`.

I suspect we can basically get away with deleting this line entirely - other than *potentially* for the note below.

Reply:
Action 2023-04-19: This is misworded: the RHS *is* evaluated, to a storage location. But the resulting storage location is not read. (So if the storage location is null - which we're going to look at or excise - we *do not* get an NRE.)

Reply:
reworded